### ISSUE TITLE
Partial overhaul of cmake + complete overhaul of cmake handling of third-party libraries

### ISSUE THREAD

**Issue by jrmadsen** [dyninst/dyninst#1100]
Hi, I am trying to utilize dyninst as a submodule and I've found that the handling of installing the 3rd-party libraries is a bit of a nightmare. 

Here is my preliminary list of changes:

- generally replace all `add_definitions`, `include_directories`, etc. with modern CMake target-based alternatives, e.g. `target_compile_definitions`, `target_include_directories`, etc. which have the capability to propagate those directives to other cmake targets which "link" to them. 
- Replace `ExternalProject_Add` with `add_subdirectory` when the subdirectory is a cmake project
- Create explicit `Dyninst_BUILD_<TPL>` options to explicitly control whether to build the third-party library or do `find_package(<TPL> REQUIRED)`
  - this has the benefit of allowing for failure if the package was not found but users do not want to build the TPL 
- Fix DyninstConfig.cmake to make a proper interface target
- Fix target installation to namespace the dyninst targets, e.g. `symtabAPI` -> `Dyninst::symtabAPI`
- In the third-party library handling, utilize `BUILD_INTERFACE` and `INSTALL_INTERFACE` generator expressions to fix the broken support for Ninja as a generator

Here are my questions: 

1. Are y'all open to accepting a PR of this nature? 
    - I'm going to do this anyway because it is completely unreliable but I just want to know if it will be accepted upstream
2. Is there any reason why y'all truly need cmake support back to v3.4.0, I consider v3.13 to be the minimal ideal minimum since it includes the `target_link_options` support

**Comment by hainest** [comment]
Yes, please! Consuming Dyninst is a bit of a nightmare right now. The original CMake system was written over ten years ago, and I've just been patching it up during my tenure. It's been on my wish list to completely rewrite it using modern CMake's target-based approach for nearly as long as I've been here, but I've just not had the time.

>generally replace all add_definitions, include_directories, etc. with modern CMake target-based alternatives, e.g. target_compile_definitions, target_include_directories, etc. which have the capability to propagate those directives to other cmake targets which "link" to them.

Agreed. There is a larger issue here in that the individual toolkits (i.e., DyninstAPI, SymtabAPI, PatchAPI, etc.) are not completely separated yet. They rely on a lot of global state to build correctly, and the third-party dependencies aren't tracked correctly (indeed, most of cmake/shared can go away). This needs to be fixed, as well.

>Replace ExternalProject_Add with add_subdirectory when the subdirectory is a cmake project

Yes, please. I have no idea why the dyninstAPI_RT toolkit build was implemented that way. It's just a pain.

>Create explicit Dyninst_BUILD_<TPL> options to explicitly control whether to build the third-party library or do find_package(<TPL> REQUIRED)

This is currently globally controlled through the STERILE_BUILD option. Ultimately, I would like to move away from building TPLs from source. It's just a maintenance hazard that offers little benefit- particularly because the bugs that show up in them are never found by users which means no one is actually using the feature. We also have plans to distribute Dyninst through container images, so this is even less of an issue.

>Fix DyninstConfig.cmake to make a proper interface target

Yup. Moving away from the train wreck that is the current `find_package(Dyninst)` would be so very beneficial to our users. For anyone reading this who isn't familiar, please see lines 4-93 of https://github.com/dyninst/testsuite/blob/master/CMakeLists.txt to see how much work it takes to use Dyninst right now.

>Fix target installation to namespace the dyninst targets, e.g. symtabAPI -> Dyninst::symtabAPI

Agreed. We should have namespaced targets for each toolkit as well as a "drag in the kitchen sink" *Dyninst::Dyninst* target.

>In the third-party library handling, utilize BUILD_INTERFACE and INSTALL_INTERFACE generator expressions to fix the broken support for Ninja as a generator

Agreed. There are many Make-specific assumptions baked into the current build system.

>Are y'all open to accepting a PR of this nature?

Yes! As I noted, this has been on my wishlist for a long time, I've just not had the cycles for it. There is a big caveat in that we'll need to preserve a lot of the existing functionality regarding TPL-specific variables, but I don't see that as a stumbling block to modernization.

>Is there any reason why y'all truly need cmake support back to v3.4.0, I consider v3.13 to be the minimal ideal minimum since it includes the target_link_options support

I believe that was the original version they used to write the current build system against. I absolutely want `target_link_options` because we need that in many places.

---

Other items:

1. get rid of the custom find modules, where possible
2. tightly constrain compiler flags between toolkits (stop polluting `CMAKE_<LANG>_FLAGS`)
3. Stop using file globs for constructing target source/header lists
4. Replace cmake/platform.cmake with the built-in CMake OS detection
5. Be more compiler-agnostic in cmake/visibility.cmake (this touches on the ugly issue of Dyninst not having a properly-defined ABI, but I don't think we will necessarily fix that here)
6. provide a formal set of user-definable options that should all go in cmake/options.cmake and mark other variables as INTERNAL so that `ccmake` users aren't inundated with a mess (as they are now).
7. clean up cmake/optimizations.cmake to be compiler-agnostic where possible (I would prefer a test-and-set methodology like what we've implemented in cmake/warnings.cmake)
8. Remove cmake/endian.cmake (this requires some additional changes to Dyninst's macro checks)
9. remove Windows checks from the build system. We don't actively support Windows, and the Windows-specific parts of Dyninst were designed for win32 not win64. We've had some folks try to build on win64, but reported errors. I think it's time for that stuff to just go. I'd need to run this up the chain, but I see no reason for it to stay.

Existing issues we should resolve during the re-write:
  #1040 - export DYNINST_INCLUDE_DIRS
  #923 - Assembly files fail to build with CMake >= 3.19 (I have an incomplete fix for this in #924)
  #707 - Canonicalize platform support  (this one is a 'maybe' as it extends beyond just the build system)

---
@kupsch @bolo Do you guys have any wishlist items for a new build system not covered here?

**Comment by jrmadsen** [comment]
@hainest Submitted PR in #1101

**Comment by jrmadsen** [comment]
With regard to #1264 and #1265 

## Features Removed

There were no features which were removed.

## Features Added

- Fatal Error is using cmake version that is too old
- Increased minimum CMake version to 3.13.0 (was 3.4.0)
- Lazy cmake install messages (i.e. `make install` only reports files which were updated)
- Message about which relevant CMake options were enabled/disabled at end of CMake run
- Quiet cmake config option when dyninst is cmake subproject
- All dyninst options can be prefixed with `DYNINST_`, e.g. `ENABLE_LTO` -> `DYNINST_ENABLE_LTO`
  - This is particularly important when Dyninst is a subproject because option names w/o prefix (such as one above) may be confusing to user when dyninst is subproject or may conflict when dyninst is subproject
- Renamed all Dyninst CMake files to have "Dyninst" prefix in their name
  - Since Dyninst installs these files, this is important to avoid filename collision
- Replaced many outdated cmake macros/functions with more portable ones
  - E.g., `add_definitions(-DSOME_DYNINST_DEF)` replaced with `add_compile_definitions(SOME_DYNINST_DEF)` which lets cmake handle what the appropriate flag is
- Use cmake INTERFACE libraries to provide flags, libs, etc. for various features
  - E.g., simply adding `Dyninst::dynCapArchDef` to `target_link_libraries(...)` provides the compile definitions
- Namespaced all targets in build-tree and install-tree
- Dramatically simplified using `find_package(Dyninst)`
  - No more loading the cache and having to add `${Boost_LIBRARIES}`, `${TBB_LIBRARIES}`, etc. to all targets -- it is simply `Dyninst::Dyninst`.
- Stable and Improved Third-party dependency installation
  - All tpls are installed to `lib/dyninst` to prevent overwriting existing installations of these deps
  - No longer install directly to the install tree. This is problematic when the install prefix requires sudo write capabilities
    - The build folders are effectively the same, e.g. Boost is in `${PROJECT_BINARY_DIR}/boost`, etc.
    - As part of the build, all TPL "install" to `${PROJECT_BINARY_DIR}/tpls` and then the contents of that directory are installed to `$<INSTALL_PREFIX>/lib/dyninst`
    - All dyninst libraries include the RPATH `$ORIGIN/dyninst/lib` to ensure any shared libraries particularly, `libdw.so`, `libelf.so`, etc., are found instead of anything within LD_LIBRARY_PATH.
    - Enabled building PIC static libraries where possible. Biggest benefit here is Boost.
    - TPLs such as TBB, elfutils are built with `$ORIGIN` RPATH to ensure that the libraries such as `libtbbmalloc_proxy.so` loads the `libtbbmalloc.so` in that directory, not another `libtbbmalloc.so` (which may be outdated) that might be installed on the system.
    - Boost source code is fetched from the Boost GitHub, preventing issues with unzipping seen occasionally on Linux
- Support for packaging via CPack
  - This allows creation of dyninst DEB, RPM, STGZ, etc. installers
  - In combination with the updates to 3rd-party deps, dyninst can now be distributed as a completely standalone distribution
    - E.g., STGZ is a shell script which contains the tarball of the pre-built installation, e.g. `./Dyninst-12.2.0-Linux.sh --prefix=/opt/dyninst --exclude-subdir --skip-license` would install dyninst to `/opt/dyninst` and not require the user to build/install appropriate versions of Boost, TBB, libiberty, or elfutils
  - Added a workflow which tests the build and installation of DEB, RPM, and STGZ installers
  - Added a workflow which automatically uploads the DEB, RPM, and STGZ installers to the "Assets" of a release

## Backwards Compatibility

Care was taken to retain nearly all backwards compatibility. The primary exception being the renaming of the files. Direct includes of say `optimizations.cmake` will fail but AFAICT, this really only affects internal dyninst testing pipelines. Some issues could arise because of the namespaced targets, e.g. `dyninstAPI_RT` target is now `Dyninst::dyninstAPI_RT` -- since the dyninstAPI_RT target does not exist, adding this to the `target_link_libraries` will now just do `-ldyninstAPI_RT`  instead of recognizing it as a target, preventing any transient linked libraries, flags, etc. from being propagated. However, prior to these modifications, Dyninst did a very poor job of propagating these, which is why you still had to add `${Boost_LIBRARIES}`, etc. to `target_link_libraries`. The way for downstream projects to check whether the DyninstConfig.cmake is old or new is quite simple (if necessary) -- one must simply check if the target `Dyninst::Dyninst` exists:

```cmake
find_package(Dyninst REQUIRED
    COMPONENTS dyninstAPI parseAPI instructionAPI symtabAPI)

if(NOT TARGET Dyninst::Dyninst) # old Dyninst CMake system was found
    # whatever was previously here
endif()
```

## Testing Done

All the [workflows in omnitrace](https://github.com/AMDResearch/omnitrace/tree/main/.github/workflows) use the updates in multiple different forms. The systems tested are:

- Ubuntu 18.04
- Ubuntu 20.04
- OpenSUSE 15.2
- OpenSUSE 15.3

Basically even version of GCC >= 7 gets tested. Some of workflows use containers where I installed Dyninst previously, e.g. `ubuntu-focal-external`, others build Dyninst as a submodule when building omnitrace.

**Comment by hainest** [comment]
>All dyninst options can be prefixed with DYNINST_, e.g. ENABLE_LTO -> DYNINST_ENABLE_LTO
>
>    This is particularly important when Dyninst is a subproject because option names w/o prefix (such as one above) may be confusing to user when dyninst is subproject or may conflict when dyninst is subproject

As far as we know, omnitrace is the first to consume dyninst as a CMake subproject so we're trying to better understand how the conflicts you mentioned arise. If a project has

    set(FOO "Hello")
    add_subdirectory(Dyninst)

and in ./Dyninst there is

    option(FOO "A FOO" "World")
    if(FOO STREQUAL "World")
      message(STATUS "case 1")
    else()
      message(STATUS "case 2")
    endif()

which case is executed here? Is there a difference if the parent project uses `set(FOO "Hello" CACHE)` or `option(FOO "" "Hello")`?

How are these cases handled for built-in variables like CMAKE_BUILD_SHARED? For example, if I want to have shared libs in the parent project but static libs from Dyninst.

Ultimately, we do not want two build parameter interfaces. If prefixed names are unavoidable, we'd rather just switch to using only them.

**Comment by jrmadsen** [comment]
> Is there a difference if the parent project uses set(FOO "Hello" CACHE) or option(FOO "" "Hello")?

No. In CMake project mode, option is basically just shorthand for setting a cache variable.

> we're trying to better understand how the conflicts you mentioned arise.

The issue, in the scenario above, is that the user will never be able to get dyninst to print `"case 1"` without modifying the cmake source, even if these users passes `cmake -D FOO=World ...`. Local variables _always_ override cache variables. It effectively makes it impossible for a user to change a variable in one project and not change the variable in another project without editing the cmake source or the upstream project having implement work-arounds. It is akin to project A defining a `DEBUG_MSG` macro in their headers and project B defining a `DEBUG_MSG` in their headers and some source file including them both. It just causes headaches which can be easily avoided by prefixing with the project name.

Here is another scenario: imagine, my project adds two subprojects, Dyninst and Foo, both Dyninst and Foo have options called `BUILD_BOOST`. How can the user get Dyninst to build Boost and Foo not to build Boost? Well, the parent project, might do something like this:

```cmake
option(BUILD_BOOST "Build boost" OFF)

add_subdirectory(Dyninst)

set(BUILD_BOOST OFF)
add_subdirectory(Foo)
```

but what if Dyninst and Foo are optional? Well then I have to:

```cmake
option(BUILD_BOOST "Build boost" OFF)

if(ENABLE_DYNINST)
    add_subdirectory(Dyninst)
endif()

if(ENABLE_FOO)
    if(ENABLE_DYNINST)
        set(BUILD_BOOST OFF)
    endif()
    add_subdirectory(Foo)
endif()
```

Now consider that Dyninst has an option `BOOST_DOWNLOAD_VERSION` which is marked as advanced. Foo has an option `BOOST_BUILD_VERSION` but they do not mark it as advanced. The user runs ccmake or cmake-gui and sees `BUILD_BOOST=ON` and `BOOST_BUILD_VERSION=1.71.0` but it keeps downloading and trying to build Boost 1.67.0, which causes errors or something. All that can easily be mitigated by those options being prefixed with `DYNINST_` and `FOO_`.

> Ultimately, we do not want two build parameter interfaces. If prefixed names are unavoidable, we'd rather just switch to using only them.

I avoided breaking as many things as possible and the implementation does this, it removes the possibility of option name conflicts while allowing all the existing cmake code to reference the variables w/o the `DYNINST_` prefix.

**Comment by jrmadsen** [comment]
> How are these cases handled for built-in variables like CMAKE_BUILD_SHARED? For example, if I want to have shared libs in the parent project but static libs from Dyninst.

Dyninst could easily handle them with:

```cmake
dyninst_add_option(BUILD_SHARED_LIBS "Build shared libs" ON)
dyninst_add_option(CMAKE_POSITION_INDEPENDENT_CODE "Build with -fPIC" ON)
```

When dyninst is the top-level project, this creates the cmake options: `BUILD_SHARED_LIBS` and `CMAKE_POSITION_INDEPENDENT_CODE` configurable by the user as usual.

When Dyninst is a sub-project, it creates cache variables `DYNINST_BUILD_SHARED_LIBS` and `DYNINST_CMAKE_POSITION_INDEPENDENT_CODE`, which can be configured by the user specifically for dyninst and because of how `dyninst_add_option` works, it overrides `BUILD_SHARED_LIBS` and `CMAKE_POSITION_INDEPENDENT_CODE` with a local variables within the project.

But the caveat here is that there are options -- such as these two options in particular -- that you probably do not want to make separately configurable from the parent project by the user, e.g. if the user specifies `-DBUILD_SHARED_LIBS=OFF -DCMAKE_POSITION_INDEPENDENT_CODE=ON`, you probably don't want to make them also provide `-DDYNINST_BUILD_SHARED_LIBS=OFF -DDYNINST_CMAKE_POSITION_INDEPENDENT_CODE=ON` when dyninst is a subproject.

A better example might be something like `dyninst_add_option(CMAKE_UNITY_BUILD "Enable unity build" OFF)`.
